packages/api/src/controllers/animalGroupController.js

```diff
@@ -0,0 +1,112 @@
+/*
+ *  Copyright 2024 LiteFarm.org
+ *  This file is part of LiteFarm.
+ *
+ *  LiteFarm is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  LiteFarm is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+
+import { Model, transaction } from 'objection';
+import baseController from './baseController.js';
+import AnimalGroupModel from '../models/animalGroupModel.js';
+import AnimalGroupRelationshipModel from '../models/animalGroupRelationshipModel.js';
+import AnimalBatchGroupRelationshipModel from '../models/animalBatchGroupRelationshipModel.js';
+import AnimalModel from '../models/animalModel.js';
+
+const animalGroupController = {
+  addAnimalGroup() {
+    return async (req, res) => {
+      const trx = await transaction.start(Model.knex());
+      try {
+        const { farm_id } = req.headers;
+        const { related_animal_ids, related_batch_ids } = req.body;
+        let { name } = req.body;
+        name = baseController.checkAndTrimString(name);
+
+        if (!name) {
+          await trx.rollback();
+          return res.status(400).send('Group name is required');
+        }
+
+        if (!Array.isArray(related_animal_ids) || !Array.isArray(related_batch_ids)) {
+          await trx.rollback();
+          return res.status(400).send('Animal ids and batch ids must be arrays');
+        }
+
+        if (!related_animal_ids.length && !related_batch_ids.length) {
+          await trx.rollback();
+          return res.status(400).send('Either animal ids or batch ids are required');
+        }
+
+        // Check ownership of animals
+        for (const animalId of related_animal_ids) {
+          const animal = await AnimalModel.query(trx)
+            .findById(animalId)
+            .where({ farm_id })
+            .whereNotDeleted();
+
+          // QUESTION: do we prefer the whole transaction to fail and an error response if any of the animals in the provided array are invalid (as written here)? Or should they be skipped and the others still inserted?
```

Comment:
In the individual animal controller, I did the first (everything fails if there's something wrong with one animal). But I think we discussed it with @SayakaOno in my PR and the second option would actually be better, create (in this case add to the group) what we can, and return in the error which entitites/records failed. I didn't pursue that with individual animals because I wanted to move on for the time being, but I think if possible and if there's not a lot of extra effort involved the second option would be preferrable

Reply:
I was about to clarify whether we would to be sending an error code or a success code in this case (I would have thought still sending a 201 for the inserted records would make more sense) but then I Googled it and realized there is an actual response status for this case, mixed error and success! Cool! [207 Multi-Status](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/207).

Reply:
I actually thought the all-or-nothing approach was good @antsgar, but when it fails, the client may want all errors depending on what's it for (eg. file upload). Partial success requires more complexed UI, so I prefer to disscuss with Loïc and Kevin before finalizing it!

Reply:
> when it fails, the client may want all errors depending on what's it for (eg. file upload)

This is a good point and I can definitely make that change (returning the array of all failures instead of the id of the first failure).